test/test_torch.py

```diff
@@ -13451,8 +13326,126 @@ def add_neg_dim_tests():
         assert not hasattr(_TestTorchMixin, test_name), "Duplicated test name: " + test_name
         setattr(_TestTorchMixin, test_name, make_neg_dim_test(name, tensor_arg, arg_constr, types, extra_dim))
 
-add_neg_dim_tests()
+# Device-agnostic tests. Instantiated below and not run directly.
+class TestTorchDeviceType(TestCase):
+    def test_diagonal(self, device):
+        x = torch.randn((100, 100), device=device)
+        result = torch.diagonal(x)
+        expected = torch.diag(x)
+        self.assertEqual(result, expected)
+
+        x = torch.randn((100, 100), device=device)
+        result = torch.diagonal(x, 17)
+        expected = torch.diag(x, 17)
+        self.assertEqual(result, expected)
+
+    def test_neg(self, device):
+        int_types = [torch.int, torch.short, torch.int8, torch.uint8]
+        float_types = [torch.float, torch.double, torch.long]
+
+        # Tests bool tensor negation raises the correct error
+        self.assertRaisesRegex(
+            RuntimeError,
+            r"Negation, the `\-` operator, on a bool tensor is not supported. "
+            r"If you are trying to invert a mask, use the `\~` or `logical_not\(\)` operator instead.",
+            lambda: - torch.tensor([False, True], device=device))
+
+        for dtype in float_types + int_types:
+            if dtype in float_types:
+                a = torch.randn(100, 90).type(dtype).to(device)
+            else:
+                a = torch.randint(-128, 128, (100, 90), dtype=dtype, device=device)
+            zeros = torch.Tensor().type(dtype).resize_as_(a).zero_().to(device)
+
+            if dtype == torch.uint8:
+                res_add = torch.add(zeros, a, alpha=255)
+            else:
+                res_add = torch.add(zeros, a, alpha=-1)
+
+            res_neg = a.clone()
+            res_neg.neg_()
+            self.assertEqual(res_neg, res_add)
+
+            # test out of place as well
+            res_neg_out_place = a.clone().neg()
+            self.assertEqual(res_neg_out_place, res_add)
+
+            # test via __neg__ operator
+            res_neg_op = -a.clone()
+            self.assertEqual(res_neg_op, res_add)
+
+    @skipCUDAIfNoMagma
```

Comment:
can I parameterize these based on device?

Reply:
Sorry, what do you mean?

Reply:
I mean I'd rather write something like:
`@skipIf(device == 'CUDA' and not TEST_MAGMA)`
than
`@skipCUDAIfNotMagma`

I'm just concerned we are going to have a lot of these decorators when they are all easily expressed with code snippets (assuming we can parameterize based on device).

Reply:
Oh I see. Yes we could support syntax like that. You'd have to put the reason in, too:

@skipIf(device == 'cuda' and not TEST_MAGMA, "no MAGMA library detected")

In practice I don't think there are many device-specific decorators, though.